### PR TITLE
feat(batch/jobs): jobs emulation

### LINKED ISSUE #44: Support for Jobs emulation
# Support for Jobs emulation

## Table of Contents

- [Support for Jobs emulation](#support-for-jobs-emulation)
  - [Table of Contents](#table-of-contents)
  - [Summary](#summary)
    - [Goals](#goals)
  - [Proposal](#proposal)
    - [Example](#example)
    - [Solution](#solution)
    - [Mappings](#mappings)
  - [Use-case](#use-case)

## Summary

Kubernetes Jobs are vital for handling one-off tasks, periodic jobs, and other similar scenarios where a task needs to run to completion before being considered successful.

Implementing support for Kubernetes Jobs will involve parsing and translating these API calls into Docker instructions, similar to how other Kubernetes resources are handled by k2d.

The base implementation could be spawning containers that will immediately reach _Running_ state, and based on the exit code (`EC`), can be marked as _Completed_ (`EC == 0`) or _Failed_ (`EC != 0`).

If you are open to considering this feature request, I would be happy to provide further input or assist in any way that I can. Please let me know if you have any questions or if there's anything I can do to support the development of this enhancement.


### Goals

1. Implementation of translation layer from Kubernetes [`jobs.batch`](https://kubernetes.io/docs/concepts/workloads/controllers/job/) to Docker containers;
2. Support for `Create`, `Delete`, `Get`, `List` and `Patch` operations;

## Proposal

### Example

The following is an example Job config (source: [kubernetes.io: Running an example job](https://kubernetes.io/docs/concepts/workloads/controllers/job/#running-an-example-job)). It computes π to 2000 places and prints it out.

```yaml
# job.yaml
apiVersion: batch/v1
kind: Job
metadata:
  name: pi
spec:
  template:
    spec:
      containers:
        - name: pi
          image: perl:5.34.0
          command: ["perl",  "-Mbignum=bpi", "-wle", "print bpi(2000)"]
      restartPolicy: Never
  backoffLimit: 4
```

Using the `kubectl create -f job.yaml`, I expect the following actions to occur:
1. **Job Creation**: When you execute `kubectl create -f job.yaml`, Kubernetes will receive this YAML configuration and create a Job resource based on the specifications defined in the YAML file.
2. **Job Scheduling**: Kubernetes will attempt to schedule the Job for execution. This involves finding an appropriate node in the cluster where the Job's pod can run. The Job pod will consist of a single container defined in the `containers` section of the Job configuration.
3. **Pod Creation**: Once a suitable node is found, Kubernetes will create a pod based on the template specified in the Job configuration. In this case, it's a single pod with a container named "pi," running the "perl:5.34.0" image.
4. **Container Initialization**: The container within the pod is started. It will execute the command specified in the `command` field of the container definition. In this example, it runs a Perl script that calculates π to 2000 decimal places using the Perl `bpi` module and prints the result.
5. **Pod Execution**: The container will run until it completes the assigned task. In this example, it computes π to 2000 decimal places and prints it. Once the command finishes, the container will terminate.
6. **Job Status Tracking**: Kubernetes continuously monitors the status of the pod and the container within the pod. If the container exits with a status code of 0 (indicating success), Kubernetes marks the pod as "Succeeded." If the container exits with a non-zero status code (indicating failure), Kubernetes marks the pod as "Failed."
7. **Job Completion**: The Job itself will be marked as "Completed" if all its pods complete successfully (i.e., with an exit code of 0) or "Failed" if any of its pods fail (i.e., exit with a non-zero code). The `backoffLimit` field in the Job configuration specifies the number of retries allowed for failed pods. In this example, it's set to 4, so Kubernetes will retry running the Job up to four times if it fails.
8. **Job Cleanup**: Once the Job is completed (either successfully or after reaching the `backoffLimit`), Kubernetes cleans up any associated resources, such as pods, unless you specify otherwise.

### Solution

From the k2d standpoint, the best way to treat the following example would be starting new container for each of the containers in the template, similarly to the [`deployment.apps` (ref. docs.k2d.io)](https://docs.k2d.io/supported-resources-and-operations#deployment).

The inspect object of the Job container would look like this (truncated to the important fields only):

```json
[
  {
    "Id": "DATA+OMITTED",
    "Path": "perl",
    "Args": ["perl", "-Mbignum=bpi", "-wle", "print bpi(2000)"],
    "State": {
      "Status": "running",
      "Running": true,
      "Paused": false,
      "Restarting": false,
      "OOMKilled": false,
      "Dead": false,
      "Pid": 41222,
      "ExitCode": 0,
      "Error": "",
      "StartedAt": "2023-09-04T15:52:43.601028251Z",
      "FinishedAt": "0001-01-01T00:00:00Z"
    },
    "Config": {
      "Cmd": ["perl", "-Mbignum=bpi", "-wle", "print bpi(2000)"],
      "Image": "perl:5.34.0",
      "Labels": {
        "namespace.k2d.io/name": "default",
        "networking.k2d.io/network-name": "k2d-default",
        "pod.k2d.io/last-applied-configuration": "DATA+OMITTED",
        "workload.k2d.io/last-applied-configuration": "DATA+OMITTED",
        "workload.k2d.io/name": "pi",
        "workload.k2d.io/type": "job"
      }
    }
  }
]
```

### Mappings

Each Pod in Kubernetes Jobs can have various statuses that indicate the progress and outcome of the job's execution, and based on them we can mark the job status. Here is a table with Possible mappings

| Job Status  | Container State | Container Exit Code       | Remarks                                             | Description                                                                                                                                                                                                                   |
|-------------|-----------------|---------------------------|-----------------------------------------------------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| `Active`    | `Running`       | -                         | -                                                   | It indicates that the Job is currently running or actively being processed. There might be one or more pods associated with the Job in the "Running" state.                                                                   |
| `Succeeded` | `Exited`        | each pod `EC == 0`        | -                                                   | It signifies that all the pods created by the Job have completed successfully. In other words, all the pods associated with the Job have exited with a status code of 0. The Job itself is considered successfully completed. |
| `Failed`    | `Exited`        | at least one pod`EC != 0` | -                                                   | It indicates that at least one pod created by the Job has failed. A pod is marked as "Failed" when its main container exits with a non-zero status code. The Job itself is considered to have failed if any of its pods fail. |
| `Completed` | `Exited`        | -                         | -                                                   | It signifies that the Job has completed its execution, but it does not provide information about whether the execution was successful or not.                                                                                 |
| `Pending`   | `Created`       | -                         | Should this be supported?                           | It means that the Job has been created, but no pods have been scheduled and started yet. Kubernetes is still in the process of scheduling the Job for execution.                                                              |
| `Unknown`   | ?               | -                         | -                                                   | It is used when there is an issue with retrieving the Job's status. This might occur if there are problems with communication between the Kubernetes control plane and the cluster nodes.                                     |
| `Expiered`  | Any state       |                           | execution time exceeds `spec.activeDeadlineSeconds` | This status is used when a Job has reached its active deadline or has been terminated due to exceeding its specified `activeDeadlineSeconds` field.                                                                           |

## Use-case

Kubernetes Jobs are primarily designed for running one-off or batch tasks in a Kubernetes cluster. They are a useful resource for managing short-lived and parallelizable workloads. Here are some common use cases for Kubernetes Jobs:

1. **Data Processing and ETL (Extract, Transform, Load)**:
   - Running data processing jobs, such as log analysis, data transformation, or data migration tasks.
   - Executing ETL pipelines to clean, enrich, and transfer data between different systems.

2. **Cron Jobs**:
   - Scheduling periodic tasks or batch jobs at specific intervals using the Kubernetes CronJob resource. CronJobs are built on top of Jobs and are suitable for tasks like periodic backups, data synchronization, or report generation.

3. **Database Migrations**:
   - Performing database schema updates or migrations as part of application deployments.
   - Running scripts to apply schema changes, seed data, or perform maintenance tasks on databases.

4. **Periodic Cleanup**:
   - Automating the cleanup of temporary files, log rotation, or purging outdated data.
   - Managing resources by periodically deleting or archiving unused objects in a cluster.

5. **Testing and CI/CD Pipelines**:
   - Running tests, build jobs, or integration tests as part of a continuous integration and continuous deployment (CI/CD) pipeline.
   - Isolating test environments and executing tests in parallel.

6. **Batch Processing**:
   - Handling parallel batch processing tasks where each unit of work is independent and can be processed concurrently.
   - Examples include processing invoices, generating reports, or transcoding media files.

7. **Backup and Restore**:
   - Automating backup and restore procedures for applications, databases, or configuration files.
   - Ensuring data resilience and disaster recovery.

8. **Resource Scaling and Parallelization**:
   - Scaling applications horizontally by launching multiple parallel instances to handle increased workloads.
   - Parallelizing tasks to complete them faster, such as image or video processing.

9. **Job Queue Workers**:
   - Creating workers to process tasks from a job queue or message queue system like RabbitMQ or Apache Kafka.
   - Scaling the number of workers based on queue backlog.

10. **Scientific and Computational Workloads**:
    - Running scientific simulations, calculations, or numerical analysis.
    - Utilizing Kubernetes clusters for high-performance computing (HPC) tasks.

11. **Backup and Restore**:
    - Automating backup and restore procedures for applications, databases, or configuration files.
    - Ensuring data resilience and disaster recovery.

12. **Ad-Hoc Administrative Tasks**:
    - Executing one-time administrative tasks or system maintenance operations on Kubernetes nodes or applications.

Kubernetes Jobs provide a reliable and declarative way to manage these tasks within a Kubernetes cluster. They ensure that tasks are executed to completion, handle retries, and provide a framework for tracking job status, which is particularly useful for monitoring and managing batch workloads.

### PR DESCRIPTION
**Pull request by shanduur**
Added support for `jobs` (API group `batch/v1`). 

Closes #44

### CODE REVIEW

**Comment by shanduur** [comment]
~~Stuck with the get API, due to the following error:~~

```log
Error from server (InternalError): an error on the server ("unable to get table: no table handler registered for this type *v1.JobList") has prevented the request from succeeding (get jobs.batch)
```

~~Error comes from [kubernetes@1.27.3/pkg/printers/tablegenerator.go:L82](https://github.com/kubernetes/kubernetes/blob/release-1.27/pkg/printers/tablegenerator.go#L82).~~

**Already fixed.**

**Comment by deviantony** [comment]
@shanduur I just saw your previous comment, this is because you're trying to display a `v1.JobList` resource as a table, you'll need to find the internal/core version of the object and use that. See other references in the code for this example.

**Comment by shanduur** [comment]
Ah yes, got this fixed! Thanks.

**Comment by shanduur** [comment]
Generally, starting jobs and listing them "works". However, the exit status of Pod is Unknown, and now the the table output looks like this:

```
$ kubectl get jobs
NAME   COMPLETIONS   DURATION   AGE
pi     0/1                      55m
```

I need to find a way to access `Status.ExitCode`, `Status.StartedAt` and `Status.FinishedAt` to correctly fill the Job response. Do you have any ideas how to handle that nicely?

My first idea was to extend [`converter.ContainerConfiguration`](https://github.com/portainer/k2d/blob/b93129295dbeeda40a8cc2e89edbd30d11a49f52/internal/adapter/converter/converter.go#L26), but I am not sure about that, and wanted to discuss it before continuing the work.

**Comment by deviantony** [comment]
@shanduur 
The original idea behind the converter.ContainerConfiguration struct was a wrapper struct that can be passed around in the context of the creation of a container so I don't think this would be the best place to store any form of status. I haven't looked at the implementation yet but have you tried to do live queries to the Docker engine (containerList, containerInspect) to retrieve the status information that you are looking for?

**Comment by shanduur** [comment]
I can access those necessary information using `container inspect`, but I haven't found a way to gracefully integrate that in the converter.

File: internal/adapter/job.go [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+	err := json.Unmarshal([]byte(jobData), &versionedJob)
+	if err != nil {
+		return nil, fmt.Errorf("unable to unmarshal versioned job: %w", err)
+	}
+
+	job := batch.Job{}
+	err = adapter.ConvertK8SResource(&versionedJob, &job)
+	if err != nil {
+		return nil, fmt.Errorf("unable to convert versioned job spec to internal job spec: %w", err)
+	}
```
Comment (shanduur) [inline internal/adapter/job.go:147]: Stuck on this line... I am able to start the job, but not list them, as it looks like there is no conversion func for `batchv1.Job` to `batch.Job`.

Comment (shanduur) [reply internal/adapter/job.go:147]:
```log
2023-09-30T01:07:27.917+0200    DEBUG   middleware/logging.go:16        received HTTP request   {"url": "/apis/batch/v1/namespaces/default/jobs?limit=500", "method": "GET", "remote_address": "192.168.0.94:55322", "request_id": "ebb1e8e9-dc4c-4447-b334-5d8454fe4f70", "header_accept": "application/json;as=Table;v=v1;g=meta.k8s.io,application/json;as=Table;v=v1beta1;g=meta.k8s.io,application/json"}
2023-09-30T01:07:27.921+0200    ERROR   utils/response.go:43    unable to get table: unable to list jobs: unable to get job: unable to convert versioned job spec to internal job spec: converting (v1.JobSpec) to (batch.JobSpec): unknown conversion      {"request_id": "ebb1e8e9-dc4c-4447-b334-5d8454fe4f70"}
github.com/portainer/k2d/internal/api/utils.ListResources
        /Users/shanduur/Repos/k2d/internal/api/utils/response.go:43
github.com/portainer/k2d/internal/api/apis/batch/jobs.JobService.ListJobs
        /Users/shanduur/Repos/k2d/internal/api/apis/batch/jobs/list.go:14
github.com/emicklei/go-restful/v3.(*FilterChain).ProcessFilter
        /Users/shanduur/Repos/k2d/vendor/github.com/emicklei/go-restful/v3/filter.go:23
main.main.CheckAuthenticationHeader.func2
        /Users/shanduur/Repos/k2d/internal/middleware/authentication.go:25
github.com/emicklei/go-restful/v3.(*FilterChain).ProcessFilter
        /Users/shanduur/Repos/k2d/vendor/github.com/emicklei/go-restful/v3/filter.go:21
github.com/portainer/k2d/internal/middleware.LogRequests
        /Users/shanduur/Repos/k2d/internal/middleware/logging.go:24
github.com/emicklei/go-restful/v3.(*FilterChain).ProcessFilter
        /Users/shanduur/Repos/k2d/vendor/github.com/emicklei/go-restful/v3/filter.go:21
github.com/portainer/k2d/internal/middleware.AddTracingHeaders
        /Users/shanduur/Repos/k2d/internal/middleware/tracing.go:14
github.com/emicklei/go-restful/v3.(*FilterChain).ProcessFilter
        /Users/shanduur/Repos/k2d/vendor/github.com/emicklei/go-restful/v3/filter.go:21
main.main.func1
        /Users/shanduur/Repos/k2d/cmd/k2d.go:156
github.com/emicklei/go-restful/v3.(*FilterChain).ProcessFilter
        /Users/shanduur/Repos/k2d/vendor/github.com/emicklei/go-restful/v3/filter.go:21
github.com/emicklei/go-restful/v3.(*Container).dispatch
        /Users/shanduur/Repos/k2d/vendor/github.com/emicklei/go-restful/v3/container.go:296
net/http.HandlerFunc.ServeHTTP
        /opt/homebrew/Cellar/go/1.21.1/libexec/src/net/http/server.go:2136
net/http.(*ServeMux).ServeHTTP
        /opt/homebrew/Cellar/go/1.21.1/libexec/src/net/http/server.go:2514
github.com/emicklei/go-restful/v3.(*Container).ServeHTTP
        /Users/shanduur/Repos/k2d/vendor/github.com/emicklei/go-restful/v3/container.go:316
net/http.serverHandler.ServeHTTP
        /opt/homebrew/Cellar/go/1.21.1/libexec/src/net/http/server.go:2938
net/http.initALPNRequest.ServeHTTP
        /opt/homebrew/Cellar/go/1.21.1/libexec/src/net/http/server.go:3546
net/http.(*http2serverConn).runHandler
        /opt/homebrew/Cellar/go/1.21.1/libexec/src/net/http/h2_bundle.go:6131
```

Comment (shanduur) [reply internal/adapter/job.go:147]: Fixed, caused by missing adapter Schema extension after rebase.

**Comment by shanduur** [comment]
Looks like the implementation is now complete... Probably?

I tested it with the following commands:

```sh
kubectl get nodes

for i in 1 2 3; do
kubectl get jobs,pods

kubectl apply -f - <<EOF
apiVersion: batch/v1
kind: Job
metadata:
  name: pi
spec:
  template:
    spec:
      containers:
        - name: pi-${i}
          image: perl:5.34.0
          command: ["perl",  "-Mbignum=bpi", "-wle", "print bpi(2000)"]
      restartPolicy: Never
  backoffLimit: 4
EOF

sleep 5

done

kubectl get jobs,pods

kubectl patch job pi -p '{"spec":{"template":{"spec":{"containers":[{"name":"pi","image":"perl:5.38.0"}]}}}}'

kubectl get jobs,pods

kubectl delete jobs pi

kubectl get jobs,pods
```

---

**EDIT:**

On the sidenote, the `Expiered` state is never reached, as I have no idea how to implement the `spec.activeDeadlineSeconds` field from the Job. This would require some additional periodical checking mechanism, or separate goroutine that could monitor the Container state and stop/kill it when Context deadline is exceeded?

**Comment by deviantony** [comment]
@shanduur thanks ! we'll be reviewing this after the current beta test phase is over.

**Comment by shanduur** [comment]
@deviantony any progres on the review? 😁

**Comment by terricain** [comment]
@deviantony any idea when this'll be merged?

**Comment by deviantony** [comment]
New releases of k2d are currently on hold. I'll keep you posted when we have more time to review it.

**Comment by deviantony** [comment]
We'll be merging this as experimental ! Thanks again @shanduur and sorry for not being available at that time.

**Comment by shanduur** [comment]
Fire! Great to see some movement in the project again!

**Pull request merged**